### PR TITLE
Add an API method to check if a payment request has been paid

### PR DESCRIPTION
**Pull request by dpad85**
This fixes https://github.com/ACINQ/eclair/issues/270
Feature was also requested by @NicolasDorier 

When `LocalPaymentHandler` receives a valid payment, it saves the payment into a `paymentDB` sqlite database.

With two new methods `allpayments` and `payment`, a user can now check through the JSON RPC API that a payment has been successfully received by a node.

For example, on node@localhost:8081, to check that payment `08d47d5f7164d4b696e8f6b62a03094d4f1c65f16e9d7b11c4a98854707e55cf` has been received, run the following curl:

```shell
curl -X POST \
  http://localhost:8081/ \
  -H 'cache-control: no-cache' \
  -H 'content-type: application/json' \
  -d '{
	"jsonrpc": "2.0",
	"method": "payment",
	"params": ["08d47d5f7164d4b696e8f6b62a03094d4f1c65f16e9d7b11c4a98854707e55cf"]
}'
```

If the payment is found, then the payment has been received. If the payment is not found, then the node has not received the payment.

---

**Note:** In the local context of a node, it is safe to consider a `paymentHash` as a unique identifier.

### CODE REVIEW

File: eclair-core/src/main/scala/fr/acinq/eclair/api/Service.scala [hunk 1]
```diff
@@ -18,27 +19,29 @@ import fr.acinq.bitcoin.{BinaryData, MilliSatoshi, Satoshi}
...
-/**
-  * Created by PM on 25/01/2016.
-  */
-
 // @formatter:off
-case class JsonRPCBody(jsonrpc: String = "1.0", id: String = "scala-client", method: String, params: Seq[JValue])
+case class JsonRPCBody(jsonrpc: String = "1.0", id: String = "eclair-node", method: String, params: Seq[JValue])
+
 case class Error(code: Int, message: String)
+
```
Comment (pm47) [inline eclair-core/src/main/scala/fr/acinq/eclair/api/Service.scala]: superfluous line break

File: eclair-core/src/main/scala/fr/acinq/eclair/api/Service.scala [hunk 2]
```diff
@@ -132,6 +136,17 @@ trait Service extends Logging {
                   sendToChannel(identifier, CMD_CLOSE(scriptPubKey = Some(scriptPubKey))).mapTo[String]
                 case JsonRPCBody(_, _, "close", JString(identifier) :: Nil) =>
                   sendToChannel(identifier, CMD_CLOSE(scriptPubKey = None)).mapTo[String]
+                case JsonRPCBody(_, _, "checkpayment", JString(identifier) :: Nil) =>
+                  for {
+                    paymentHash <- Try(PaymentRequest.read(identifier)) match {
+                        case Success(pr) => Future.successful(pr.paymentHash)
+                        case _ => Try(BinaryData(identifier)) match {
+                          case Success(s) => Future.successful(s)
+                          case _ => Future.failed(new IllegalArgumentException("payment identifier must be a Payment Request or a Payment Hash"))
```
Comment (pm47) [inline eclair-core/src/main/scala/fr/acinq/eclair/api/Service.scala]: inconsistent error message

File: eclair-core/src/main/scala/fr/acinq/eclair/api/Service.scala [hunk 3]
```diff
@@ -148,6 +163,8 @@ trait Service extends Logging {
                     "send (paymentRequest, amountMsat): send a payment to a lightning node using a BOLT11 payment request and a custom amount",
                     "close (channelId): close a channel",
                     "close (channelId, scriptPubKey): close a channel and send the funds to the given scriptPubKey",
+                    "allpayments: list all received payments",
```
Comment (pm47) [inline eclair-core/src/main/scala/fr/acinq/eclair/api/Service.scala]: remove `allpayments`

File: eclair-core/src/main/scala/fr/acinq/eclair/api/Service.scala [hunk 4]
```diff
@@ -148,6 +163,8 @@ trait Service extends Logging {
...
+                    "payment (paymentHash or paymentRequest): returns the payment if it has been received",
```
Comment (pm47) [inline eclair-core/src/main/scala/fr/acinq/eclair/api/Service.scala]: maybe two lines like for `close`

```
"payment (paymentHash): returns the payment if it has been received",
"payment (paymentRequest): returns the payment if it has been received",
```

File: eclair-core/src/main/scala/fr/acinq/eclair/payment/LocalPaymentHandler.scala [hunk 5]
```diff
@@ -1,8 +1,11 @@
 package fr.acinq.eclair.payment
 
+import java.time.ZonedDateTime
```
Comment (pm47) [inline eclair-core/src/main/scala/fr/acinq/eclair/payment/LocalPaymentHandler.scala]: better use same technique as for Announcements

File: eclair-core/src/main/scala/fr/acinq/eclair/db/Payment.scala [hunk 6]
```diff
@@ -0,0 +1,12 @@
...
+import fr.acinq.bitcoin.BinaryData
+
+/**
+  * Payment object stored in DB.
+  *
+  * @param payment_hash identifier of the payment
+  * @param amount_msat amount of the payment, in milli-satoshis
+  * @param timestamp absolute time in seconds since UNIX epoch when the payment was created.
+  */
+case class Payment(payment_hash: BinaryData, amount_msat: Long, timestamp: Long)
```
Comment (pm47) [inline eclair-core/src/main/scala/fr/acinq/eclair/db/Payment.scala]: define in `object PaymentDb`

File: eclair-core/src/main/scala/fr/acinq/eclair/api/Service.scala (hunk 2 above)
Comment (dpad85) [reply eclair-core/src/main/scala/fr/acinq/eclair/api/Service.scala]: 👍

File: eclair-core/src/main/scala/fr/acinq/eclair/api/Service.scala (hunk 1 above)
Comment (dpad85) [reply eclair-core/src/main/scala/fr/acinq/eclair/api/Service.scala]: 👍

File: eclair-core/src/main/scala/fr/acinq/eclair/api/Service.scala (hunk 4 above)
Comment (dpad85) [reply eclair-core/src/main/scala/fr/acinq/eclair/api/Service.scala]: 👍

File: eclair-core/src/main/scala/fr/acinq/eclair/api/Service.scala (hunk 3 above)
Comment (dpad85) [reply eclair-core/src/main/scala/fr/acinq/eclair/api/Service.scala]: 👍

File: eclair-core/src/main/scala/fr/acinq/eclair/db/Payment.scala (hunk 6 above)
Comment (dpad85) [reply eclair-core/src/main/scala/fr/acinq/eclair/db/Payment.scala]: 👍

**Comment by NicolasDorier** [comment]
awesome! I saw that a bit late. But another very useful thing would be a long poll method that I can call, which returns when a new payment happen or after timeout. Lighting-C has this, and this is very useful:

If there is 10 simultaneous invoices about to be paid, I don't have to poll 10 request per seconds to know what has been paid. Instead I can just call the long-poll 'waitforpayment' then fetch with `payment` method.

**Comment by AndySchroder** [comment]
Any way to return the value paid instead of true or false? For the case where you provide a payment request without a specific amount, there doesn't seem to be a way to know how much they actually paid.

**Pull request merged**